Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt

```diff
@@ -216,16 +159,313 @@ class DefaultRiskLevels @Inject constructor() : RiskLevels {
+        val mostRecentDateWithLowRisk = mostRecentDateForRisk(
+            exposureHistory,
+            RiskCalculationParametersOuterClass.NormalizedTimeToRiskLevelMapping.RiskLevel.LOW
+        )
+
+        Timber.d("mostRecentDateWithLowRisk: $mostRecentDateWithLowRisk")
+
+        // 8. Determine `Date of Most Recent Date with High Risk`
+        val mostRecentDateWithHighRisk = mostRecentDateForRisk(
+            exposureHistory,
+            RiskCalculationParametersOuterClass.NormalizedTimeToRiskLevelMapping.RiskLevel.HIGH
+        )
+
+        Timber.d("mostRecentDateWithHighRisk: $mostRecentDateWithHighRisk")
+
+        // 9. Determine `Total Minimum Distinct Encounters With Low Risk`
+        val totalMinimumDistinctEncountersWithLowRisk = exposureHistory
+            .sumBy { it.minimumDistinctEncountersWithLowRisk }
+
+        Timber.d("totalMinimumDistinctEncountersWithLowRisk: $totalMinimumDistinctEncountersWithLowRisk")
+
+        // 10. Determine `Total Minimum Distinct Encounters With High Risk`
+        val totalMinimumDistinctEncountersWithHighRisk = exposureHistory
+            .sumBy { it.minimumDistinctEncountersWithHighRisk }
+
+        Timber.d("totalMinimumDistinctEncountersWithHighRisk: $totalMinimumDistinctEncountersWithHighRisk")
+
+        return AggregatedRiskResult(
+            totalRiskLevel,
+            totalMinimumDistinctEncountersWithLowRisk,
+            totalMinimumDistinctEncountersWithHighRisk,
+            mostRecentDateWithLowRisk,
+            mostRecentDateWithHighRisk
+        )
+    }
+
+    private fun mostRecentDateForRisk(
+        exposureHistory: List<AggregatedRiskPerDateResult>,
+        riskLevel: RiskCalculationParametersOuterClass.NormalizedTimeToRiskLevelMapping.RiskLevel
+    ): Instant? = exposureHistory
+        .filter { it.riskLevel == riskLevel }
+        .maxOfOrNull { it.dateMillisSinceEpoch }
+        ?.let { Instant.ofEpochMilli(it) }
+
+    private fun aggregateRiskPerDate(
+        dateMillisSinceEpoch: Long,
+        exposureWindowsAndResult: Map<ExposureWindow, RiskResult>
+    ): AggregatedRiskPerDateResult {
+        // 1. Group `Exposure Windows by Date`
+        val exposureWindowsAndResultForDate = exposureWindowsAndResult
+            .filter { it.key.dateMillisSinceEpoch == dateMillisSinceEpoch }
+
+        // 2. Determine `Normalized Time per Date`
+        val normalizedTime = exposureWindowsAndResultForDate.values
+            .sumOf { it.normalizedTime }
+
+        Timber.d("Aggregating result for date $dateMillisSinceEpoch - ${Instant.ofEpochMilli(dateMillisSinceEpoch)}")
+
+        // 3. Determine `Risk Level per Date`
+        val riskLevel = try {
+            appConfig.normalizedTimePerDayToRiskLevelMappingList
+                .filter { it.normalizedTimeRange.inRange(normalizedTime) }
+                .map { it.riskLevel }
+                .first()
+        } catch (e: Exception) {
+            throw ApplicationConfigurationInvalidException(
```

Comment:
Maybe this should be checked by the config mapper. If the config is invalid if it doesn't contain it, and we throw during mapping (parsing), then the Config logic has the chance to provide an alternative config from storage or default.